src/Control/Controller.php

```diff
@@ -680,6 +680,18 @@ public static function join_links($arg = null)
      */
     public static function normaliseTrailingSlash(string $url): string
     {
+        // Do not normalise external urls
+        // Note that urls without a scheme such as www.example.com will be counted as a relative file
+        if (!Director::is_site_url($url)) {
+            return $url;
+        }
+
+        // Do not modify files e.g. some/path/myfile.txt (exclude if $url is absolute_base_url)
+        $extension = pathinfo(rtrim($url), PATHINFO_EXTENSION);
+        if ($extension && $url !== Director::config()->get('alternate_base_url')) {
```

Comment:
```suggestion
        if (preg_match('/^(.*)\.([^\/]*)$/', Director::makeRelative($url))) {
```
This seems to work the way we'd expect this to work, and doesn't have the problem in https://github.com/silverstripe/silverstripe-framework/pull/11321#discussion_r1704908697
This regex was what was originally being used to detect files.

This would probably work as well:
```suggestion
        $extension = pathinfo(Director::makeRelative($url), PATHINFO_EXTENSION);
        if ($extension && $url !== Director::config()->get('alternate_base_url')) {
```

Reply:
Updated